go_2-dev: init at dev.go2go branch

3noch opened the pull request:
###### Motivation for this change

Allow people to use the go2 proposal easily from Nix.

I'm not familiar with the build for Go so I only did things that managed to get this branch to build. ccing Go team in case they want to help support this cause in a more official way: @c00w @cstrahan @Frostman @kalbasit @mdlayher @mic92 @orivej @rvolosatovs @zowoq 

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [ ] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

SuperSandro2000 commented on pkgs/development/compilers/go/2-dev.nix:
```diff
@@ -0,0 +1,262 @@
+
+  checkPhase = ''
+    runHook preCheck
+    (cd src && HOME=$TMPDIR GOCACHE=$TMPDIR/go-cache ./run.bash --no-rebuild)
```
```suggestion

   pushd src
   HOME=$TMPDIR GOCACHE=$TMPDIR/go-cache ./run.bash --no-rebuild
   popd

```

SuperSandro2000 commented on pkgs/top-level/all-packages.nix:
```diff
@@ -9525,6 +9525,13 @@ in
   });
 
+  go_2-dev = callPackage ../development/compilers/go/2-dev.nix ({
+    inherit (darwin.apple_sdk.frameworks) Security Foundation;
```
```suggestion
    inherit (darwin.apple_sdk.frameworks) Foundation Security;
```

zowoq commented:
@SuperSandro2000 I'd rather that this remain a copy of the existing `go1.*.nix` files for easier diffing, formatting changes can/should be applied to all of them.

3noch commented:
@SuperSandro2000 Yes this is just a copy of other go.nix files. Happy to clean all of them up but that should be a separate PR I'd imagine. I can separate out the maintainer changes. I'm probably not a good *choice* of maintainer for this since I don't actively use Go any more.

3noch commented on pkgs/development/compilers/go/2-dev.nix:
```diff
@@ -0,0 +1,262 @@
+  '';
+
+  postBuild = ''
+    (cd src && ./make.bash)
```
I actually prefer the original way here. But like I said this is a copy.

c00w commented:
Looks fine to me - I vote to merge :) (Pedantically I might make a go 2 folder for this, but honestly the go folder despite being a bit messy has never gotten in my way, so I don't worry about it too much).

3noch committed:
maintainers: add _3noch

3noch committed:
go_2-dev: init at dev.go2go branch

Head of dev.go2go branch was abe4d3dce12252ed09216eaa67b7dab8c8922537
at time of writing.

3noch commented on maintainers/maintainer-list.nix:
```diff
@@ -70,6 +70,12 @@
     githubId = 7414843;
     name = "Nicholas von Klitzing";
   };
+  _3noch = {
```
Done.

kalbasit commented:
@GrahamcOfBorg build go_2-dev

kalbasit commented:
The commit 

> go: Add dev.go2go branch variant called 2-dev

Does not really follow CONTRIBUTING.md and should be something like this: `go_2-dev: init at ...`

Otherwise LGTM. I'll merge once the commit is fixed and the bot builds it.

3noch commented:
@kalbasit How's that?

kalbasit commented:
nix-review result

```
1 package built:
go_2-dev

[nix-shell:~/.cache/nixpkgs-review/pr-107538-1]$ ./results/go_2-dev/bin/go version
go version 2 linux/amd64
```

fishy commented:
With the merge of this (or at least I believe so, I'm new to nixpkgs), my `go-1.15.6` auto upgraded to `go-2` with `nix-channel --update && nix-env -u`.

Should we rename this one as something like `go2-2` to avoid that? Or what's the correct way to "pin" my go at `1.*`, but still keep updated?

zowoq commented:
> Should we rename this one as something like `go2-2` to avoid that?

#108270

3noch commented:
Yikes! I had no idea the update functionality worked like that. So thankful you pointed that out.

kalbasit commented:
> Yikes! I had no idea the update functionality worked like that. So thankful you pointed that out.

Me neither :scream: Nix has some really weird edge cases.
